### ISSUE TITLE
Gravity-influenced entangled blocks can become 'out-of-sync' if falling

### ISSUE THREAD

**Issue by stanleyrya** [TeacherGaming/qcraft-mod#9]
First mentioned here: TeacherGaming/qcraft-mod#5

If a gravity-influenced entangled block is falling, it will not resolve with it's entangled group to any other block type. This is noticeable when you resolve it to a non-gravity-influenced entangled block while falling and it remains a gravity-influenced entangled block. The 'out of sync' entangled block, once on the ground, will return to being in sync upon the next new observation.

Steps to reproduce:

1. Create a new creative mode world in single player.
2. Grab 4 blocks of a non-gravity-influenced block (stone), 4 blocks of a gravity-influenced block (sand), two Essence of Superposition, and one Essence of Entanglement.
3. Create two quantum blocks using the above ingredients, then use the Essence of Entanglement to entangle them.
4. Place a few of the entangled quantum blocks on the ground.
5. Using some block, build directly up next to your entangled quantum blocks until you are at a very high height.
6. Add an entangled quantum block at the top.
7. Make sure it is resolved to a non-gravity-influenced block and put on your Anti-Observation Goggles.
8. Remove all blocks that are below your high entangled quantum block.
9. Remove your Anti-Observation Goggles and resolve the entangled quantum blocks on the ground to a gravity-influenced block, then QUICKLY resolve them to a non-gravity-influenced block before the high entangled quantum block hits the ground.
10. Notice how the once high entangled quantum block remains a gravity-influenced block, even after hitting the ground.
11. When you resolve the blocks another time, everything returns to normal.

It is also possible to refrain from fixing this 'bug' and keeping it on as a feature. Would resolving a block to non-gravity-influenced while in-between two block positions cause a crash?

**Comment by stanleyrya** [comment]
Before falling:  
<img width="1440" alt="screen shot 2016-02-18 at 3 57 36 pm" src="https://cloud.githubusercontent.com/assets/8700100/13158827/cda03402-d65c-11e5-9d22-21e07acd12a1.png">

Starting the fall:  
<img width="1440" alt="screen shot 2016-02-18 at 3 57 46 pm" src="https://cloud.githubusercontent.com/assets/8700100/13158823/cd9dc6e0-d65c-11e5-873d-f3f5faed4f0a.png">

Observed as non-gravity-influenced while falling:  
<img width="1440" alt="screen shot 2016-02-18 at 3 57 52 pm" src="https://cloud.githubusercontent.com/assets/8700100/13158825/cd9df656-d65c-11e5-8bc8-46bc35a69277.png">

Observed again to get back in sync:  
<img width="1440" alt="screen shot 2016-02-18 at 3 58 00 pm" src="https://cloud.githubusercontent.com/assets/8700100/13158824/cd9e0592-d65c-11e5-9c1e-023334a1eaf1.png">
<img width="1440" alt="screen shot 2016-02-18 at 3 58 10 pm" src="https://cloud.githubusercontent.com/assets/8700100/13158826/cd9eff4c-d65c-11e5-9a27-94fb9bf73519.png">

**Comment by Robijnvogel** [comment]
So how should we go about fixing this?
My validate() addition as proposed in the #5 pull request broke a lot of other stuff.
The place where I added this validate(), however, is probably where some other code should be added. 

These 2 things should be established whenever a falling QBlock hits the ground and becomes a block again:
1. It should be re-connected to the entanglement group it was/is a part of.
2. It should change to the state that the rest of the blocks in this entanglement group are in.

**Comment by stanleyrya** [comment]
While I'm not opposed to fixing this issue, I think we should consider leaving it as it is. It seems to be a relatively niche issue and at least with my testing it even fixes itself upon next observation. We should also consider the possible added code complexity that changing the current behavior could have. Unfortunately this isn't something I know too much about; it is just my first time contributing to a Minecraft mod.

I guess the biggest thing to think about would be if there is a solid use case for fixing this behavior. For example, does leaving this behavior in cause any game-breaking bugs? Or does it get in the way of certain building structures/traps and limits the capabilities of the mod?

**Comment by Robijnvogel** [comment]
If this really fixes itself (actually entangles itself), like you say it does, then I guess the glitch/bug is not that bad.
If it doesn't fix itself and doesn't re-entangle to the other blocks (And in your tests, just behaves that way because it is so near to the other blocks that you have always looked at it from the same side as the other blocks, forcing it into the same block state.), it should be fixed, because that would make it almost useless in continuous machines. 
It may still be useful in some one-time trap, but it can not be re-used.

We should test which of these two things is really happening though.

**Comment by Robijnvogel** [comment]
So I've done some digging around and found out why my fix that worked in OpenMC's branch breaks things right now:

In OpenMC's branch, the "entanglementregistry" reset every time you quit and then rejoined a world.
In this official version that is already fixed.
My "validate()" addition gets called as well on game restart. In OpenMC's branch this actually fixed the reset on game restart.
In this official branch, my "validate()" doubly registers every entangled block to the entanglementregistry, which breaks it.

Now what we have to find is a piece of code that only gets called on the solidification of FallingQBlocks and not on game-start. Adding validate(); or some other form of EntanglementRegistry.register(int, this) there should fix it without breaking other things.

I wonder if the blocks, before they start falling, get unregistered from the EntanglementRegistry at all.

**Comment by Robijnvogel** [comment]
https://youtu.be/F29wZgj5BI0
I guess it does not fix itself.
EDIT: With my "validate();" fix I have observed what you described, but that broke other stuff...

**Comment by Robijnvogel** [comment]
I do not get how my validate(); could break other stuff.

The [super.validate();](https://github.com/TeacherGaming/qcraft-mod/blob/master/src/main/java/dan200/qcraft/shared/TileEntityQBlock.java#L110) or [m_nextUnusedFrequency](https://github.com/TeacherGaming/qcraft-mod/blob/master/src/main/java/dan200/qcraft/shared/EntanglementRegistry.java#L70) are not what breaks it as I have tested without them and it still breaks.

For as far as I can see, this only leaves the [m_entanglements.get( frequency ).add( entangledObject );](https://github.com/TeacherGaming/qcraft-mod/blob/master/src/main/java/dan200/qcraft/shared/EntanglementRegistry.java#L73) , but that should not be able to cause an issue, as the way that ArrayLists work, they can not add an object that they already contain.

**Comment by Robijnvogel** [comment]
So I've found the onStopFalling method in the Minecraft source code. It's not deobfuscated, but using a simple System.out.println function I've ascertained myself that this is, in fact, what I thought it was.

However, overriding the method and implementing a call of the validate() function on the world.getTileEntity(x, y, z) TileEntity does not seem to influence anything.

Maybe this is because the onStopFalling "event" gets triggered just a little too soon for the validate method to have effect?

This deobfuscated method in my development environment is: func_149828_a at line 116 in Blockfalling.java .
